test/old_tests/UnitTests/apartment_context.cpp

```diff
@@ -56,6 +56,64 @@ namespace
 
         REQUIRE(is_nta_on_mta());
     }
+
+    IAsyncAction TestStaToStaApartmentContext()
+    {
+        bool pass = false;
+
+        apartment_context original;
+
+        // Create an STA thread and switch to it.
+        auto controller1 = DispatcherQueueController::CreateOnDedicatedThread();
+        co_await resume_foreground(controller1.DispatcherQueue());
+
+        // Save the COM context for the first STA thread.
+        apartment_context context1;
+
+        // Create another STA thread and switch to it.
+        auto controller2 = DispatcherQueueController::CreateOnDedicatedThread();
+        co_await resume_foreground(controller2.DispatcherQueue());
+
+        // The first STA thread remains hung as long as this is true.
+        bool hang_sta_1 = true;
+
+        // Hang the first STA thread.
+        controller1.DispatcherQueue().TryEnqueue([&] {
+            bool hang = true;
+            while (hang_sta_1)
+            {
+                if (!WaitOnAddress(&hang_sta_1, &hang, sizeof(hang), 1000))
+                {
+                    return; // failed - timed out
+                }
+            }
+            pass = true;
+        });
+
+        // Queue work to the second STA thread to unstick the first thread.
+        // This requires that the second STA thread be available to dispatch work
+        // after the "co_await context1" below.
+        controller2.DispatcherQueue().TryEnqueue([&] {
+            hang_sta_1 = false;
+            WakeByAddressAll(&hang_sta_1);
+        });
+
+        // Try to switch to the first STA thread via COM context.
+        // The first STA thread is hung, but once this thread (the second STA thread)
+        // processes work, it will unstick the first STA thread.
+        // This test verifies that the second STA thread does become available
+        // to do work and is not hung waiting for the first STA thread.
+        co_await context1;
```

Comment:
So if this were broken (aka before your changes) this co_await would block forever if not for the `1000` timeout on line 85.  Is that right?

Reply:
Right. My initial version had `INFINITE`, and the test hung. We don't like it when tests hang, so I had it detect that it had been stuck for 1 second (should be plenty) and declare failure if so.